docs/architecture/adr-108-e2e-abci++.md

```diff
@@ -0,0 +1,305 @@
+logged by this node. Remember here that `fetchABCIRequests()` returns an array of slices(`[]*abci.Request`) where the slice 
+with index 0 corresponds to the node's `CleanStart` execution, and each additional slice corresponds to the `Recovery` 
+execution after a specific crash. Each node must have one `CleanStart` execution and the same number of `Recovery` executions 
+as the number of crashes that happened on this node. If collecting was successful, the test checks whether each execution 
+respects the ABCI++ 
+grammar by calling `checker.Verify()` method. If `Verify` returns an error, the specific execution does not respect the 
+grammar, and the test will fail. 
+
+The tests are executed only if `ABCITestsEnabled` is set to `true`. This is done through the manifest file. Namely, if we 
+want to test whether CometBFT respects ABCI++ grammar, we would need to enable these tests by adding `abci_tests_enabled = 
+true` in the manifest file of a particular testnet (e.g. `networks/ci.toml`). This will automatically activate logging on the 
+application side. 
+
+The `Verify()` method is shown below. 
+```go
+func (g *GrammarChecker) Verify(reqs []*abci.Request, isCleanStart bool) (bool, error) {
+	if len(reqs) == 0 {
+		return false, fmt.Errorf("execution with no ABCI calls.")
+	}
+	r := g.filterRequests(reqs)
+	// Check if the execution is incomplete.
+	if len(r) == 0 {
+		return true, nil
+	}
+	var errors []*Error
+	execution := g.getExecutionString(r)
+	if isCleanStart {
+		errors = g.verifyCleanStart(execution)
+	} else {
+		errors = g.verifyRecovery(execution)
+	}
+	if errors == nil {
+		return true, nil
+	}
+	return false, fmt.Errorf("%v\nFull execution:\n%v", g.combineErrors(errors, g.cfg.NumberOfErrorsToShow), g.addHeightNumbersToTheExecution(execution))
+}
+```
+
+It receives a set of ABCI++ requests and a flag saying whether they represent a `CleanStart` execution or not and does the following things:
+- Checks if the execution is an empty execution. 
+- Filter the requests by calling the method `filterRequests()`. This method will remove all the requests from the set that are not supported by the current version of the grammar. In addition, it will filter the last height by removing all ABCI++ requests after the 
+last `Commit`. The function `fetchABCIRequests()` can be called in the middle of the height. As a result, the last height may be incomplete and 
+classified as invalid, even if that is not the reality. The simple example here is that the last 
+request fetched via `fetchABCIRequests()` is `FinalizeBlock`; however, `Commit` happens after 
+`fetchABCIRequests()` was invoked. Consequently, the execution
+will be considered as faulty because `Commit` is missing, even though the `Commit` 
+will happen after. This is why if the execution consists of only one incomplete height and function `filterRequests()` returns an empty set of requests, the `Verify()` method considers this execution as valid and returns `true`. 
```

Comment:
If could be good to mention here that this mechanism effectively turns the original grammar, which is infinite (omega grammar) into a finite one. In terms of the grammar, we change multiplier `(inf)` by `*`.

Anyway, this is a nice-to-have. Up to you if you want to explain it here.

Reply:
Addressed. Not precisely here, but in the part where we explain the difference between the new and original grammar.